docs/config.md
# Onionbalance configuration tool

## Description

The `onionbalance-config` tool is the fastest way to generate the
necessary keys and config files to get your onion service up and
running.

```bash
onionbalance-config
```

When called without any arguments, the config generator will run in an
interactive mode and prompt for user input.

The `config` directory should be stored on the management server.

## Files

Onionbalance requires the following files:

* `config.yaml`: this is the configuration file that is used by the
  Onionbalance management server.

* `ONION_ADDRESS.key`: Each service instance have a private key, which
   will be derived into the public address and identity for the Onion Service.
   Each private key file is named after the .onion address, so
   `ONION_ADDRESS.key` will be actually like
   `dpkhemrbs3oiv2fww5sxs6r2uybczwijzfn2ezy2osaj7iox7kl7nhad.key`.
   It is essential that you keep this file secure.

By default, onionbalance search for these files in the current directory
where it was invoked.

The actual folder can be explicitly set by invoking `onionbalance-config` with
the `--output` option.

When running the system-wide service, this folder is automatically set
to `/etc/onionbalance`.

## Environment variables

<!-- This behavior was not ported to v3 Onion Services -->
<!--
The loaded configuration file takes precedence over environment variables.
Configuration file options will override environment variable which have the
same name.
-->

* `ONIONBALANCE_CONFIG`: Override the location for the Onionbalance
  configuration file.

* `ONIONBALANCE_LOG_LEVEL`: Specify the minimum verbosity of log messages to
  output. All log messages equal or higher the specified log level are
  output. The available log levels are the same as the `--verbosity` command
  line option.

* `ONIONBALANCE_STATUS_SOCKET_LOCATION`: The Onionbalance service creates a
  Unix domain socket which provides real-time information about the currently
  loaded service and descriptors. This option can be used to change the
  location of this domain socket. (default: `/var/run/onionbalance/control`).

* `ONIONBALANCE_TOR_CONTROL_SOCKET`: The location of the Tor unix domain
  control socket. Onionbalance will attempt to connect to this control socket
  first before falling back to using a control port connection. (default:
  `/var/run/tor/control`).


docs/onionbalance-config.md
# Configuration tool

## Description

The `onionbalance-config` tool is the fastest way to generate the
necessary keys and config files to get your onion service up and
running.

```console
$ onionbalance-config
```

When called without any arguments, the config generator will run in an
interactive mode and prompt for user input.

The `master` directory should be stored on the management server while
the other `instance` directories should be transferred to the respective
backend servers.

## Files

* `master/config.yaml`: This is the configuration file that is used my the
  Onionbalance management server.

* `master/<ONION_ADDRESS>.key`: The private key which will become the public
  address and identity for your onion service. It is essential that you keep
  this key secure.

* `master/torrc-server`: A sample Tor configuration file which can be used with
  the Tor instance running on the management server (v2-only).

* `srv/torrc-instance`: A sample Tor config file which contains the Tor
  `HiddenService*` options needed for your backend Tor instance (v2-only).

* `srv/<ONION_ADDRESS>/private_key`: Directory containing the private key for
  your backend onion service instance. This key is less critical as it can be
  rotated if lost or compromised (v2-only).


CHANGES.rst
0.2.2
-----

- Add an OBv3 hacking guide.
- Remove tox and simplify build procedure.
- A single OnionBalance can now support multiple onion services.

0.2.1
-----

- v2 codebase now uses Cryptodome instead of the deprecated PyCrypto library.
- v3 codebase is now more flexible when it comes to requiring a live
  consensus. This should increase the reachability of Onionbalance in scenarios
  where the network is having trouble establishing a new consensus.
- v3 support for connecting to the control port through a Unix socket. Patch by Peter Tripp.
- Introduce status socket support for v3 onions. Patch by vporton.
- Sending a SIGHUP signal now reloads the v3 config. Patch by Peter Chung.

0.2.0
-----

- Allow migration from Tor to Onionbalance by reading tor private keys directly
  using the 'key' directive in the YAML config file. Also update
  `onionbalance-config` to support that.
- Improve `onionbalance-config` for v3 onions. Simplify the output directory
  (and change docs to reflect so) and the wizard suggestions.

0.1.9
-----

- Initial support for v3 onions!

0.1.8
-----

- Fix a bug which could cause descriptor fetching to crash and stall if an
  old instance descriptor was retrieved from a HSDir. #64
- Minors fixes to documentation and addition of a tutorial.

0.1.7
-----

- Add functionality to reconnect to the Tor control port while Onionbalance is
  running. Thank you to Ceysun Sucu for the patch. #45
- Fix bug where instance descriptors were not updated correctly when an
  instance address was listed under multiple master service. #49
- Improve performance by only requesting each unique instance descriptor
  once per round, rather once for each time it was listed in the config
  file. #51
- Fix bug where an exception was raised when the status socket location did
  not exist.
- Improve the installation documentation for Debian and Fedora/EPEL
  installations.

0.1.6
-----

- Remove unicode tags from the yaml files generated by onionbalance-config.
- Fix bug resulting in invalid instance onion addresses when attempting to
  remove the ".onion" TLD. #44

0.1.5
-----

- Log error when Onionbalance does not have permission to read a private key. #34
- Fix bug loading descriptors when an address with .onion extension is listed
  in the configuration file. #37
- Add support for connecting to the Tor control port over a unix domain socket. #3

0.1.4
-----

- Use setproctitle to set a cleaner process title
- Replace the python-schedule dependency with a custom scheduler.
- Add a Unix domain socket which outputs the status of the Onionbalance
  service when a client connects. By default this socket is created at
  `/var/run/onionbalance/control`. Thank you to Federico Ceratto for the
  original socket implementation.
- Add support for handling the `SIGINT` and `SIGTERM` signals. Thank you to
  Federico Ceratto for this feature.
- Upgrade tests to use the stable Tor 0.2.7.x release.
- Fix bug when validating the modulus length of a provided RSA private key.
- Upload distinct service descriptors to each hidden service directory by
  default. The distinct descriptors allows up to 60 introduction points or
  backend instances to be reachable by external clients. Thank you to Ceysun
  Sucu for describing this technique in his Masters thesis.
- Add `INITIAL_DELAY` option to wait longer before initial descriptor
  publication. This is useful when there are many backend instance descriptors
  which need to be downloaded.
- Add configuration option to allow connecting to a Tor control port on a
  different host.
- Remove external image assets when documentation is generated locally
  instead of on ReadTheDocs.

0.1.3
-----

- Streamline the integration tests by using Tor and Chutney from the
  upstream repositories.
- Fix bug when HSFETCH is called with a HSDir argument (3d225fd).
- Remove the 'schedule' package from the source code and re-add it as a
  dependency. This Python package is now packaged for Debian.
- Extensively restructure the documentation to make it more comprehensible.
- Add --version argument to the command line
- Add configuration options to output log entries to a log file.

0.1.2
-----

- Remove dependency on the schedule package to prepare for packaging
  Onionbalance in Debian. The schedule code is now included directly in
  onionbalance/schedule.py.
- Fix the executable path in the help messages for onionbalance and
  onionbalance-config.

0.1.1
-----

- Patch to resolve issue when saving generated torrc files from
  onionbalance-config in Python 2.


0.1.0
-----

-  Initial release


onionbalance/hs_v2/config.py
# -*- coding: utf-8 -*-
import os

"""
Define default config options for the management server
"""

# Set default configuration options for the management server

REPLICAS = 2
HSDIR_SET = 3  # Publish each descriptor to 3 consecutive HSDirs
MAX_INTRO_POINTS = 10
DESCRIPTOR_OVERLAP_PERIOD = 60 * 60
DESCRIPTOR_UPLOAD_PERIOD = 60 * 60  # Re-upload descriptor every hour
REFRESH_INTERVAL = 10 * 60
PUBLISH_CHECK_INTERVAL = 5 * 60
INITIAL_DELAY = 45  # Wait for instance descriptors before publishing

LOG_LOCATION = os.environ.get('ONIONBALANCE_LOG_LOCATION')
LOG_LEVEL = os.environ.get('ONIONBALANCE_LOG_LEVEL', 'info')

STATUS_SOCKET_LOCATION = os.environ.get('ONIONBALANCE_STATUS_SOCKET_LOCATION',
                                        '/var/run/onionbalance/control')
TOR_CONTROL_PASSWORD = None

# Upload multiple distinct descriptors containing different subsets of
# the available introduction points
DISTINCT_DESCRIPTORS = True

# Store global data about onion services and their instance nodes.
services = []

controller = None
